### DISCUSSION TITLE
issue: reset not working with useLayoutEffect

### DISCUSSION THREAD

**Discussion by pavelspichonak** [react-hook-form/react-hook-form#8888] [General]
### Version Number

7.33.1

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-reset-v7-forked-ru52zz?file=/src/index.js

### Steps to reproduce

1. use **reset** method inside **useLayoutEffect** hook

### Expected behaviour

**reset** method must update form values like it does when it is used inside **useEffect** hook

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
Thanks for the issue report, reset API is required to have all the input reference finished registration with actual DOM, which `useLayoutEffect` to stay synchronously performing all the DOM manipulation.  I would recommend using `useEffect` to reset API instead. let me know your specific use case, maybe I can suggest a workaround instead.

**Comment by pavelspichonak** [comment]
@bluebill1049 I have a form, for example user profile. I use **isDiry** flag to be able to show preventing browser modal when browser tab is closed. User can change form values and move to another page(without saving form values to backend) so I store form values for example to redux store.

Then user can come to user profile page again, so I pass original values(values from backend) to **defaultValues** and need to use **reset(currentValues, { keepDefaultValues: true })** to set current form values from redux store to be able to know if form is dirty. All works good but if I use **reset** inside **useEffect** hook user for a moment sees original form values(from backend) and then after milliseconds user sees current values from redux store. That is why **useLayoutEffect** is necessary for my case.

It is not critical but I believe all of us are perfectionists and such visible changing form values is annoying  :)

**Reply by bluebill1049** [reply]
I see what you mean now, maybe that's the case need for a loading phase before displaying the form with values?

**Reply by pavelspichonak** [reply]
I'm reading documentation one more time, maybe I can use current values(from redux) as **defaultValues** and in **useEffect** I can use original values from backend like this **reset(defaultValues, { keepValues: true })**? Looks like in this case user will not see visible changes. I'll try. The main idea is to use **keepValues** instead of **keepDefaultValues**

**Reply by bluebill1049** [reply]
there is also another option which is keepDirtyValues, keep user modified data, but you will have to subscribe to `dirtyFields`.

**Reply by pavelspichonak** [reply]
Yeah, my option to use **keepValues** instead of **keepDefaultValues** works but **isDirty** is false until user interaction with form.

**Reply by pavelspichonak** [reply]
**currentValues** - values from redux store
**defaultValues** - initial values from backend

So final solution is on form initialisation I use
```
useForm({
    defaultValues: currentValues,
  });
```
and **useEffect** looks like this:
```
useEffect(() => {
    reset(defaultValues);
    reset(currentValues, { keepDefaultValues: true });
  }, []);
```
too complicated but looks like it works

**Comment by leapful** [comment]
@pavelspichonak @bluebill1049 : Actually, it can be simplified by just perform reset with default values (from BE) on "useLayoutEffect" and then update current values (from redux) on "useEffect".

Since in phase "useLayoutEffect", resetting with default values only updates default values in react-hook-form, no DOM mutation, so you will not get into a flash on DOM input element and then after phase "useEffect", the value is render on DOM input element just one time.

```
useLayoutEffect(() => {
    reset(defaultValues);
  }, [reset]);

  useEffect(() => {
    reset(currentValues, { keepDefaultValues: true });
  }, [reset]);
```

Working demo: https://codesandbox.io/s/react-hook-form-reset-v7-forked-93ymbw?file=/src/index.js

**Reply by pavelspichonak** [reply]
@leapful Thanks. It works. But what about **defaultValues** passed to **useForm**? **useForm** requires all fields in **defaultValues** to be present. Is it okay to not pass **defaultValues** to **useForm**? Is it expected behaviour that we have form inputs/selects etc but do not have **defaultValues** passed to **useForm**?

> It is encouraged that you set defaultValues for all inputs to non-undefined values such as the empty string or null.

**Reply by leapful** [reply]
@pavelspichonak : Method "reset" will update default values just like passing it to "useForm". The tricky part is that "useForm" is a hook and will apply default values on next rendering while calling "reset" in phase "useLayoutEffect" performs update default values at the time before actual DOM input is registered therefore it will not cause blinking/flashing default value and current value on UI.
